pkg/providers/tinkerbell/tinkerbell.go

```diff
@@ -448,6 +461,102 @@ func (p *tinkerbellProvider) setMachinesToPXEBoot(ctx context.Context) error {
 	return errorutil.NewAggregate(errs)
 }
 
+// scrubWorkflowsFromTinkerbell removes all workflows in the Tinkerbell stack that feature in hardware by retrieving
+// hardware MAC addresses using tinkerbellHardware. tinkerbellHardware is necessary because MAC addresses aren't
+// available on the Hardware object type.
+func (p *tinkerbellProvider) scrubWorkflowsFromTinkerbell(ctx context.Context, hardware []tinkv1alpha1.Hardware, tinkerbellHardware []*tinkhardware.Hardware) error {
+	workflows, err := p.providerTinkClient.GetWorkflow(ctx)
+	if err != nil {
+		return fmt.Errorf("retrieving workflows: %w", err)
+	}
+
+	hardwareMACLookup, err := createHardwareIDToMACMapping(tinkerbellHardware)
+	if err != nil {
+		return err
+	}
+
+	manifestHardwareMACs, err := createMACSetFromHardwareManifests(hardwareMACLookup, hardware)
+	if err != nil {
+		return err
+	}
+
+	workflowIDs, err := getWorkflowsIDsFromMACs(manifestHardwareMACs, workflows)
+	if err != nil {
+		return err
+	}
+
+	if err := p.providerTinkClient.DeleteWorkflow(ctx, workflowIDs...); err != nil {
+		return fmt.Errorf("could not delete tinkerbell workflow: %v", err)
+	}
+
+	return nil
+}
+
+func createHardwareIDToMACMapping(hardware []*tinkhardware.Hardware) (map[string]string, error) {
+	hardwareMACLookup := make(map[string]string)
+	for _, h := range hardware {
+		if len(h.Network.Interfaces) == 0 {
+			return nil, fmt.Errorf("hardware manifest without interface: hardware ID = '%v'", h.Id)
+		}
+		hardwareMACLookup[h.Id] = h.Network.Interfaces[0].Dhcp.Mac
+	}
+
+	return hardwareMACLookup, nil
+}
+
+func createMACSetFromHardwareManifests(hardwareMACLookup map[string]string, hardware []tinkv1alpha1.Hardware) (macAddressSet, error) {
+	manifestHardwareMACs := make(macAddressSet)
+	for _, h := range hardware {
+		mac, found := hardwareMACLookup[h.Spec.ID]
+		if !found {
+			return nil, fmt.Errorf("couldn't find mac address for hardware manifest: manifest hardware ID = '%v'", h.Spec.ID)
+		}
+
+		manifestHardwareMACs.Insert(mac)
+	}
+
+	return manifestHardwareMACs, nil
+}
+
+func getWorkflowsIDsFromMACs(hardwareMACs macAddressSet, workflows []*tinkworkflow.Workflow) ([]string, error) {
+	var workflowIDs []string
+	for _, w := range workflows {
+		mac, err := macFromWorkflow(w)
+		if err != nil {
+			return nil, err
```

Comment:
this would just return the error message that unmarshal returns. it would be good to add more specific things to this error here. we might also discuss later if we should print an error and continue with the other workflows here instead of completely bailing

Reply:
Do you mean adding some context?